cnd/cnd.lsp/src/org/netbeans/modules/cnd/lsp/compilationdb/ClangCDBSupport.java

```diff
@@ -0,0 +1,219 @@
+        LOG.log(Level.FINE, "File operations finished.");
+    }
+
+    @Override
+    public void configurationXmlChanged(MakeProjectEvent ev) {
+        updateCompilationDatabase(ClangCDBGenerationCause.PROJECT_CONFIGURATION_CHANGED);
+    }
+
+    /**
+     * Ignored. See javadoc
+     * @param ev 
+     */
+    @Override
+    public void propertiesChanged(MakeProjectEvent ev) {
+        LOG.log(Level.FINE, "Some project properties changed {0}",
+                ev.getPath());
+    }
+
+    @Override
+    protected void projectOpened() {
+        LOG.log(Level.FINE, "Project opened");
+        projectOpen = true;
+        makeProject.getHelper().addMakeProjectListener(this);
+        NativeProject nativeProject = makeProject.getLookup().lookup(NativeProject.class);
+        if (nativeProject != null) {
+            nativeProject.addProjectItemsListener(this);
+        } else {
+            LOG.log(Level.FINE, "Cannot attach to native project events.");
+        }
+        updateCompilationDatabase(ClangCDBGenerationCause.PROJECT_OPENED);
+    }
+
+    @Override
+    protected void projectClosed() {
+        LOG.log(Level.FINE, "Project closed");
+        removeListenersOnCloseOrDelete();
+    }
+
+    @Override
+    public void projectDeleted(NativeProject nativeProject) {
+        LOG.log(Level.FINE, "Project deleted");
+        removeListenersOnCloseOrDelete();
+    }
+
+    private void removeListenersOnCloseOrDelete() {
+        projectOpen = false;
+        makeProject.getHelper().removeMakeProjectListener(this);
+        NativeProject nativeProject = makeProject.getLookup().lookup(NativeProject.class);
+        if (nativeProject != null) {
+            nativeProject.removeProjectItemsListener(this);
+        } else {
+            LOG.log(Level.FINE, "Cannot detach from native project events.");
+        }
+    }
+
+    /**
+     * Updates the compilation database "[project]/compile_commands.json". If a
+     * task is already running then the request is silently ignored.
+     *
+     * @param cause The cause that originated the need to recreate the
+     * compilation database.
+     */
+    private void updateCompilationDatabase(ClangCDBGenerationCause cause) {
+        LOG.log(Level.FINE, "Update compilation database project is open: {0}", projectOpen);
+        if (projectOpen) {
+            pendingCauses.add(cause);
+            ClangCDBGenerationTask task = null;
+            try {
+                task = new ClangCDBGenerationTask(this, pendingCauses);
+            } catch (IllegalArgumentException iae) {
+                LOG.log(Level.SEVERE, "Unable to create ClangCDBGenerationTask: {0}", iae.getMessage());
+                return;
+            }
+            // We fire the regeneration task after a COALESCING_DELAY_MS milliseconds, to
+            // coalesce requests (for instance, when many files are added or removed).
+            CLANG_CDB_PROCESSOR.schedule(task, COALESCING_DELAY_MS, TimeUnit.MILLISECONDS);
```

Comment:
I'm not sure this is what you want. Jan made me aware of a mode of the request processor when working on the LSPBindings.  The background tasks are created: 

https://github.com/apache/netbeans/blob/48ff3e7790554fffae05eab5621a9f1271642951/ide/lsp.client/src/org/netbeans/modules/lsp/client/LSPBindings.java#L444-L457

and then scheduled to be run:

https://github.com/apache/netbeans/blob/48ff3e7790554fffae05eab5621a9f1271642951/ide/lsp.client/src/org/netbeans/modules/lsp/client/LSPBindings.java#L475-L477

The important part is here:
- if `#scheduleBackgroundTask` is called while the task is already waiting, execution is again postponed
- if the task is finished, it is scheduled again
- if the task already running it is left to be running to completion, but a new run is scheduled

https://bits.netbeans.org/12.0/javadoc/org-openide-util/org/openide/util/RequestProcessor.Task.html#schedule-int-

For the LSPBindings this means, that the updates in NetBeans happen, when the situation is settled. This might be a sane approach here too.